### PR TITLE
Add Windows LV2 support (EXPERIMENTAL)

### PR DESCRIPTION
**Pull request by AnClark**
By adding corresponding LV2 UI code, SynthV1 is available on Windows now. Windows LV2 hosts (e.g. REAPER v6.28) could load SynthV1's Qt UI. 

image

**KNOWN ISSUE:**
- Plugin will crash if you delete the SynthV1 instance from REAPER's FX list, then re-add it. Backtrace from GDB shows that crash point is in `Qt6Gui.dll`.
- CMake only. I didn't improve Autoconf files, as Autotools runs extremely slow on Msys2.

### CODE REVIEW

**Comment by rncbc** [comment]
awesome, I think :)

first question: why have a separate `src/synthv1_windows.lv2/` for the *.ttl ? couldn't the new entries be merged into the existing common `src/synthv1.lv2/`*.ttl ? this way it'll be kind of an undesirable burden to maintain the whole stuff.

the crashes are probably a symptom to way reaper or windows deals with the Qt .dll's; have you tried to build and link to Qt static libraries instead?

**Comment by AnClark** [comment]
> first question: why have a separate src/synthv1_windows.lv2/ for the .ttl ?

Currently `src/synthv1.lv2/` is for Linux only. Merging them together is OK, but a tiny preprocessor is required to generate TTL files for those two platforms. 

I've got an idea: use `sed` to preprocess TTL templates, just like this implementation in [setBfree](https://github.com/pantherb/setBfree/tree/master/b_synth):

```makefile
manifest.ttl: manifest.ttl.in manifest.ui.ttl.in
	sed "s/@LV2NAME@/$(LV2NAME)/;s/@LIB_EXT@/$(LIB_EXT)/;s/@MODGUITTL@/$(MODGUITTL)/" manifest.ttl.in > manifest.ttl
ifeq ($(HAVE_UI), yes)
	sed "s/@LV2NAME@/$(LV2NAME)/;s/@LV2GUI@/$(LV2GUI)/;s/@LIB_EXT@/$(LIB_EXT)/;s/@UI_TYPE@/$(UI_TYPE)/" manifest.ui.ttl.in >> manifest.ttl
endif

$(LV2NAME).ttl: $(LV2NAME).ttl.in $(LV2NAME).ui.ttl.in
	sed "s/@VERSION@/lv2:microVersion $(LV2MIC) ;lv2:minorVersion $(LV2MIN) ;/g;s/@MODBRAND@/$(MODBRAND)/;s/@MODLABEL@/$(MODLABEL)/" \
		$(LV2NAME).ttl.in > $(LV2NAME).ttl
ifeq ($(HAVE_UI), yes)
	sed "s/@UI_TYPE@/$(UI_TYPE)/;s/@UI_REQ@/$(LV2UIREQ)/;" $(LV2NAME).ui.ttl.in >> $(LV2NAME).ttl
endif
```

> the crashes are probably a symptom to way reaper or windows deals with the Qt .dll's; have you tried to build and link to Qt static libraries instead?

I've worked for two days trying to statically link SynthV1 with static Qt. But I always got `undefined references` when GCC invokes `ld`. Would you love to give me a hand?

**Comment by AnClark** [comment]
Here's my script to build static Qt on Windows (Qt base only):

```batch
cmd.exe /c "configure.bat -static -debug-and-release -platform win32-g++ -prefix $QtDir `
        -qt-zlib -qt-pcre -qt-libpng -qt-libjpeg -qt-freetype -qt-harfbuzz -sqlite -opengl desktop -no-openssl `
        -opensource -confirm-license `
        -make libs -nomake tools -nomake examples -nomake tests"
```

Before executing this command, I modified `mkspecs\win32-g++\qmake.conf` in Qt source:

```makefile
# Append this to the end of file
# [QT-STATIC-PATCH]
QMAKE_LFLAGS += -static -static-libgcc
QMAKE_CFLAGS_RELEASE -= -O2
QMAKE_CFLAGS_RELEASE += -Os -momit-leaf-frame-pointer
DEFINES += QT_STATIC_BUILD
```

**Comment by rncbc** [comment]
please move to Qt6 >= 6.1

Qt5 is no longer a target for opensource projects, no matter the platform

cheers

ps. also rebase to the "develop" branch

**Comment by AnClark** [comment]
> please move to Qt6 >= 6.1
> 
> Qt5 is no longer a target for opensource projects, no matter the platform

OK, I'll sync my local fork.

But I still cannot link SynthV1 to static Qt6.

**Comment by AnClark** [comment]
> first question: why have a separate src/synthv1_windows.lv2/ for the .ttl ?

I've done a solution: a8b9442

**Comment by AnClark** [comment]
Update 2021/5/25: Now I can build a static version of SynthV1, and it won't crash on my REAPER.

**Comment by rncbc** [comment]
yay. good news! thanks

now, what about that macro/template thing... do you feel it's really necessary?
why not just adding the lv2ui_windows entry as is? i believe it won't make no arm to any host or platform that doesn't support it anyhow (ie. Linux; but I maybe wrong, coz I won't touch anything else)

**Comment by AnClark** [comment]
Using template is a solution to your question before:

> first question: why have a separate src/synthv1_windows.lv2/ for the .ttl ? couldn't the new entries be merged into the existing common src/synthv1.lv2/.ttl ? this way it'll be kind of an undesirable burden to maintain the whole stuff.

---

I'm just new to LV2's TTL configuration file, and I'm not sure if it works well when I put entries for different platforms together in one file. By now using a template should be a better choice, and would make it easy to port SynthV1 to macOS in future.

Maybe I need a test.

**Comment by rncbc** [comment]
oh wait, yes you probably need to hack for the `lv2:binary` entry on the manifest.ttl and synthv1_ui.ttl, but everything else it's probably fine

**Comment by AnClark** [comment]
OK, I'll give it a try on PadthV1's implementation.

---

UPDATE: Only one platform-specific entry is allowed. If you put `ui_windows` and `ui_x11` together in `synthv1.ttl`, the UI won't load. So it's better to hack all of those 3 files with my method.

**Comment by rncbc** [comment]
I really think it would be a lot more simple if you have a separate `manifest.ttl` file (eg. `manifest_win32.ttl`) that just declares `lv2:binary <synthv1.dll> ;` (and not the `.so` suffix as is for Linux) and then installs/copies that file into `manifest.ttl` on deployment.

I believe having both 'ui_windows' and 'ui_x11' simultaneously anyway in `synthv1.ttl` and `synthv1_ui.ttl` is no cause for concern, or is it? at least qtractor (Linux) won't choke on that ;)

**Comment by rncbc** [comment]
fyi. this last commit on develop branch, b71df43 is actually intended for you ;)

in addition to the aforementioned `manifest-win32.ttl` , you can also add now this `synthv1_ui-win32.ttl`, that (I recommend) to have just this literal contents:
~~~
@prefix lv2:     <http://lv2plug.in/ns/lv2core#> .
@prefix lv2ui:   <http://lv2plug.in/ns/extensions/ui#> .

@prefix synthv1_lv2: <http://synthv1.sourceforge.net/lv2#> .

<http://synthv1.sourceforge.net/lv2>
	lv2ui:ui synthv1_lv2:ui_windows, synthv1_lv2:ui_external .

<http://synthv1.sourceforge.net/lv2#ui_windows>
	a lv2ui:WindowsUI ;
	lv2:requiredFeature <http://lv2plug.in/ns/ext/instance-access> ;
	lv2:optionalFeature lv2ui:resize, lv2ui:idleInterface, lv2ui:showInterface ;
	lv2:extensionData lv2ui:resize, lv2ui:idleInterface, lv2ui:showInterface ;
	lv2ui:binary <synthv1.dll> .

<http://synthv1.sourceforge.net/lv2#ui_external>
	a <http://kxstudio.sf.net/ns/lv2ext/external-ui#Widget> ;
	lv2:requiredFeature <http://lv2plug.in/ns/ext/instance-access> ;
	lv2ui:binary <synthv1.dll> .
~~~
nb. dunno whether the `ui_external` entry makes any sense on windows, but let's assume it's OK.

so that, in the install phase you shall just do this file copy/rename:
`manifest-win32.ttl` -> `manifest.ttl`
`synthv1_ui-win32.ttl` -> `synthv1_ui.ttl`

what do you think?

cheers

**Comment by rncbc** [comment]
this PR is now merged upstream as in new [develop-win32](https://github.com/rncbc/synthv1/tree/develop-win32) branch;
includes the above suggestions in https://github.com/rncbc/synthv1/pull/14#issuecomment-857860171;

please rebase to this branch in all follow-ups and communication re. this PR.#14; maybe it's one better idea to open up another PR, if you have any further fixes (just because you probably will ;))

wholly thanks && cheers

**Comment by AnClark** [comment]
Great! I've seen b71df43 (as well as the one in PadthV1). But I still don't think that splitting Win32 and Linux version to different branches is a better choice, as:
- They share all the codes (by setting different macros or options, the same code tree can fit up with different platforms);
- It would be complicated to maintain more than 1 develop branch. Every time you need to merge / cherry-pick changes into develop-win32 branch manually.

**Comment by rncbc** [comment]
> I still don't think that splitting Win32 and Linux version to different branches is a better choice

the [develop-win32](https://github.com/rncbc/synthv1/tree/develop-win32)  branch is just for you to checkout and test your stuff on windows;) eventually, when all's clear on your side, it should get merged into main [develop](https://github.com/rncbc/synthv1/tree/develop)  branch.

**Comment by AnClark** [comment]
> the develop-win32 branch is just for you to checkout and test your stuff on windows;) eventually, when all's clear on your side, it should get merged into main develop branch.

So it is. It may help me a lot.

**Comment by rncbc** [comment]
re. [README.windows.md](https://github.com/rncbc/synthv1/blob/develop-win32/README.windows.md):

TL;DR: instructions assume you're based on ArchLinux, as a cross-build platform...

is this actually intended?

**Comment by AnClark** [comment]
> TL;DR: instructions assume you're based on ArchLinux, as a cross-build platform...

Msys2 is a platform on Windows, providing useful tools to build native and Cygwin-compatible Unix applications.
This means I only build Windows version on Windows, with Msys2.
Cross-build is supported theorically, but not tested.

See: <https://www.msys2.org/>

**Comment by rncbc** [comment]
> > TL;DR: instructions assume you're based on ArchLinux, as a cross-build platform...
> 
> Msys2 is a platform on Windows, providing useful tools to build native and Cygwin-compatible Unix applications.
> This means I only build Windows version on Windows, with Msys2.
> Cross-build is supported theorically, but not tested.
> 

so what does all that `pacman` stands for? ;)

**Comment by AnClark** [comment]
Pacman is also Msys2's package manager 😋 The authors just implemented it into Msys2 so that every user can enjoy a brilliant experience for package managements.

See Msys2's introduction:

> To provide easy installation of packages and a way to keep them updated it features a package management system called Pacman, which should be familiar to Arch Linux users. It brings many powerful features such as dependency resolution and simple complete system upgrades, as well as straight-forward and reproducible package building.

**Comment by rncbc** [comment]
notes. what about adding that very information to the README.windows.md ?

**Comment by AnClark** [comment]
> what about adding that very information to the README.windows.md ?

It's OK, but may not be necessary, as I assume that users know what Msys2 is.

**Comment by rncbc** [comment]
as you might know already this has been mostly merged into develop branch today.

also to all the other vee-ones: [samplv1](https://github.com/rncbc/samplv1/tree/develop), [drumkv1](https://github.com/rncbc/drumkv1/tree/develop) and [padthv1](https://github.com/rncbc/padthv1/tree/develop) :)
thanks a lot!
cheers

**Comment by AnClark** [comment]
My pleasure!
I'll check if they work well on my environment.
